File: src/datamgr/dm_engine/dm_engine/utils/formatter.py
# -*- coding: utf-8 -*-
"""
Tencent is pleased to support the open source community by making BK-BASE 蓝鲸基础平台 available.
Copyright (C) 2021 THL A29 Limited, a Tencent company.  All rights reserved.
BK-BASE 蓝鲸基础平台 is licensed under the MIT License.
License for BK-BASE 蓝鲸基础平台:
--------------------------------------------------------------------
Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated
documentation files (the "Software"), to deal in the Software without restriction, including without limitation
the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software,
and to permit persons to whom the Software is furnished to do so, subject to the following conditions:
The above copyright notice and this permission notice shall be included in all copies or substantial
portions of the Software.
THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT
LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN
NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,
WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.

summary:
    This library is provided to allow standard python
    logging to output log data as JSON formatted strings
    ready to be shipped out to logstash.
"""

import datetime
import json
import logging
import socket
import traceback as tb


def _default_json_default(obj):
    """
    Coerce everything to strings.
    All objects representing time get output as ISO8601.
    """
    if isinstance(obj, (datetime.datetime, datetime.date, datetime.time)):
        return obj.isoformat()
    else:
        return str(obj)


class LogstashFormatter(logging.Formatter):
    """
    A custom formatter to prepare logs to be
    shipped out to logstash.
    """

    def __init__(self, fmt=None, datefmt=None, style="%", json_cls=None, json_default=_default_json_default):
        """
        :param fmt: Config as a JSON string, allowed fields;
               extra: provide extra fields always present in logs
               source_host: override source host name
        :param datefmt: Date format to use (required by logging.Formatter
            interface but not used)
        :param json_cls: JSON encoder to forward to json.dumps
        :param json_default: Default JSON representation for unknown types,
                             by default coerce everything to a string
        """

        if fmt is not None:
            self._fmt = json.loads(fmt)
        else:
            self._fmt = {}
        self.json_default = json_default
        self.json_cls = json_cls
        if "extra" not in self._fmt:
            self.defaults = {}
        else:
            self.defaults = self._fmt["extra"]
        if "source_host" in self._fmt:
            self.source_host = self._fmt["source_host"]
        else:
            try:
                self.source_host = socket.gethostname()
            except Exception:
                self.source_host = ""

    def format(self, record):
        """
        Format a log record to JSON, if the message is a dict
        assume an empty message and use the dict as additional
        fields.
        """

        fields = record.__dict__.copy()

        if isinstance(record.msg, dict):
            fields.update(record.msg)
            fields.pop("msg")
            msg = ""
        else:
            msg = record.getMessage()

        try:
            msg = msg.format(**fields)
        except (KeyError, IndexError, ValueError):
            pass
        except Exception:
            # in case we can not format the msg properly we log it as is instead of crashing
            msg = msg

        if "msg" in fields:
            fields.pop("msg")

        if "exc_info" in fields:
            if fields["exc_info"]:
                formatted = tb.format_exception(*fields["exc_info"])
                fields["exception"] = formatted
            fields.pop("exc_info")

        if "exc_text" in fields and not fields["exc_text"]:
            fields.pop("exc_text")

        logr = self.defaults.copy()

        logr.update(
            {
                "@message": msg,
                "@timestamp": datetime.datetime.utcnow().strftime("%Y-%m-%dT%H:%M:%S.%fZ"),
                "@source_host": self.source_host,
                "@fields": self._build_fields(logr, fields),
            }
        )

        return json.dumps(logr, default=self.json_default, cls=self.json_cls)

    def _build_fields(self, defaults, fields):
        """Return provided fields including any in defaults

        >>> f = LogstashFormatter()
        # Verify that ``fields`` is used
        >>> f._build_fields({}, {'foo': 'one'}) == \
                {'foo': 'one'}
        True
        # Verify that ``@fields`` in ``defaults`` is used
        >>> f._build_fields({'@fields': {'bar': 'two'}}, {'foo': 'one'}) == \
                {'foo': 'one', 'bar': 'two'}
        True
        # Verify that ``fields`` takes precedence
        >>> f._build_fields({'@fields': {'foo': 'two'}}, {'foo': 'one'}) == \
                {'foo': 'one'}
        True
        """
        return dict(list(defaults.get("@fields", {}).items()) + list(fields.items()))


class LogstashFormatterV1(LogstashFormatter):
    """
    A custom formatter to prepare logs to be
    shipped out to logstash V1 format.
    """

    def format(self, record):
        """
        Format a log record to JSON, if the message is a dict
        assume an empty message and use the dict as additional
        fields.
        """

        fields = record.__dict__.copy()

        if "msg" in fields and isinstance(fields["msg"], dict):
            msg = fields.pop("msg")
            fields.update(msg)

        elif "msg" in fields and "message" not in fields:
            msg = record.getMessage()
            # fields.pop('msg')

            try:
                msg = msg.format(**fields)
            except (KeyError, IndexError):
                pass
            except Exception:
                # in case we can not format the msg properly we log it as is instead of crashing
                msg = msg
            fields["message"] = msg

        if "exc_info" in fields:
            if fields["exc_info"]:
                formatted = tb.format_exception(*fields["exc_info"])
                fields["exception"] = formatted
            fields.pop("exc_info")

        if "exc_text" in fields and not fields["exc_text"]:
            fields.pop("exc_text")

        now = datetime.datetime.utcnow()
        base_log = {
            "@timestamp": now.strftime("%Y-%m-%dT%H:%M:%S") + ".%03d" % (now.microsecond / 1000) + "Z",
            "@version": 1,
            "source_host": self.source_host,
        }
        base_log.update(fields)

        logr = self.defaults.copy()
        logr.update(base_log)

        return json.dumps(logr, default=self.json_default, cls=self.json_cls)

File: src/dataweb/web/src/components/monaco/sql-format/core/Formatter.js
/*
 * Tencent is pleased to support the open source community by making BK-BASE 蓝鲸基础平台 available.
 * Copyright (C) 2021 THL A29 Limited, a Tencent company.  All rights reserved.
 * BK-BASE 蓝鲸基础平台 is licensed under the MIT License.
 *
 * License for BK-BASE 蓝鲸基础平台:
 * --------------------------------------------------------------------
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated
 * documentation files (the "Software"), to deal in the Software without restriction, including without limitation
 * the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software,
 * and to permit persons to whom the Software is furnished to do so, subject to the following conditions:
 * The above copyright notice and this permission notice shall be included in all copies or substantial
 * portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT
 * LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN
 * NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,
 * WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE
 */

/* eslint-disable no-param-reassign */
/* eslint-disable no-plusplus */
/* eslint-disable no-underscore-dangle */
import { trimEnd } from './utils';
import tokenTypes from './tokenTypes';
import Indentation from './Indentation';
import InlineBlock from './InlineBlock';
import Params from './Params';

export default class Formatter {
  /**
     * @param {Object} cfg
     *   @param {Object} cfg.indent
     *   @param {Object} cfg.params
     * @param {Tokenizer} tokenizer
     */
  constructor(cfg, tokenizer) {
    this.cfg = cfg || {};
    this.indentation = new Indentation(this.cfg.indent);
    this.inlineBlock = new InlineBlock();
    this.params = new Params(this.cfg.params);
    this.tokenizer = tokenizer;
    this.previousReservedWord = {};
  }

  /**
     * Formats whitespaces in a SQL string to make it easier to read.
     *
     * @param {String} query The SQL query string
     * @return {String} formatted query
     */
  format(query) {
    const tokens = this.tokenizer.tokenize(query);
    const formatTokens = this.formatTokens(tokens);
    const formattedQuery = this.getFormattedQueryFromTokens(formatTokens);

    return formattedQuery.trim();
  }

  /**
     *  根据配置每行字段个数格式化换行
     *  针对Select Update 等设置字段时
     */
  formatTokens(tokens) {
    const tokenLen = tokens.length;
    let positoin = 0;
    const _tokens = [];

    while (positoin < tokenLen) {
      const token = tokens[positoin];
      _tokens.push(token, tokens[positoin + 1]);
      if (token.type === tokenTypes.RESERVED_TOPLEVEL) {
        const __tokens = this.getNextLimitToken(tokens, positoin);
        _tokens.push(...__tokens.tokens);
        positoin = __tokens.positoin;
      } else {
        positoin++;
      }
    }

    return _tokens;
  }

  getNextLimitToken(tokens, positoin) {
    const wordLimit = this.cfg.wordLimit || 3;
    let stop = false;
    const _tokens = [];
    const returnTokens = [];
    while (!stop) {
      positoin++;
      const _token = tokens[positoin];
      stop = !_token || _token.type === tokenTypes.RESERVED_TOPLEVEL;
      !stop && _tokens.push(_token);
    }

    const wordstring = _tokens.reduce(
      (pre, next) => ((next.value === ',' && pre.trimEnd()) || pre) + next.value,
      ' ',
    );
    const wordlist = wordstring.trim().split(/\s+|\n+/);
    let start = 0;
    let end = wordLimit;

    do {
      returnTokens.push(
        { type: tokenTypes.RESERVED_NEWLINE_AFTER, value: wordlist.slice(start, end).join(' ') },
        { type: tokenTypes.WHITESPACE, value: '' },
      );
      end = end + wordLimit;
      start = start + wordLimit;
    } while (start < wordlist.length);

    return { tokens: returnTokens, positoin };
  }

  getFormattedQueryFromTokens(tokens) {
    let formattedQuery = '';

    tokens.forEach((token, index) => {
      if (!token) {
        return;
      }
      if (token.type === tokenTypes.WHITESPACE) {
        return;
      }
      if (token.type === tokenTypes.LINE_COMMENT) {
        formattedQuery = this.formatLineComment(token, formattedQuery);
      } else if (token.type === tokenTypes.BLOCK_COMMENT) {
        formattedQuery = this.formatBlockComment(token, formattedQuery);
      } else if (token.type === tokenTypes.RESERVED_TOPLEVEL) {
        formattedQuery = this.formatToplevelReservedWord(token, formattedQuery);
        this.previousReservedWord = token;
      } else if (token.type === tokenTypes.RESERVED_NEWLINE) {
        formattedQuery = this.formatNewlineReservedWord(token, formattedQuery);
        this.previousReservedWord = token;
      } else if (token.type === tokenTypes.RESERVED_NEWLINE_AFTER) {
        formattedQuery = this.formatNewlineAfterReservedWord(token, formattedQuery);
        this.previousReservedWord = token;
      } else if (token.type === tokenTypes.RESERVED) {
        formattedQuery = this.formatWithSpaces(token, formattedQuery);
        this.previousReservedWord = token;
      } else if (token.type === tokenTypes.OPEN_PAREN) {
        formattedQuery = this.formatOpeningParentheses(tokens, index, formattedQuery);
      } else if (token.type === tokenTypes.CLOSE_PAREN) {
        formattedQuery = this.formatClosingParentheses(token, formattedQuery);
      } else if (token.type === tokenTypes.PLACEHOLDER) {
        formattedQuery = this.formatPlaceholder(token, formattedQuery);
      } else if (token.value === ',') {
        formattedQuery = this.formatComma(token, formattedQuery);
      } else if (token.value === ':') {
        formattedQuery = this.formatWithSpaceAfter(token, formattedQuery);
      } else if (token.value === '.' || token.value === ';') {
        formattedQuery = this.formatWithoutSpaces(token, formattedQuery);
      } else {
        formattedQuery = this.formatWithSpaces(token, formattedQuery);
      }
    });
    return formattedQuery;
  }

  formatLineComment(token, query) {
    return this.addNewline(query + token.value);
  }

  formatBlockComment(token, query) {
    return this.addNewline(this.addNewline(query) + this.indentComment(token.value));
  }

  indentComment(comment) {
    return comment.replace(/\n/g, `\n${this.indentation.getIndent()}`);
  }

  formatToplevelReservedWord(token, query) {
    this.indentation.decreaseTopLevel();
    query = this.addNewline(query);
    this.indentation.increaseToplevel();
    query += this.equalizeWhitespace(token.value.toUpperCase());
    return `${query} `; // this.addNewline(query)
  }

  formatNewlineReservedWord(token, query) {
    return `${this.addNewline(query) + this.equalizeWhitespace(token.value)} `;
  }

  formatNewlineAfterReservedWord(token, query) {
    const isTypeEquals = this.previousReservedWord.type === tokenTypes.RESERVED_TOPLEVEL;
    const isAfterTopLevel = this.previousReservedWord && isTypeEquals;
    return `${(isAfterTopLevel ? query : this.addNewline(query)) + token.value} `;
  }

  // Replace any sequence of whitespace characters with single space
  equalizeWhitespace(string) {
    return string.replace(/\s+/g, ' ');
  }

  // Opening parentheses increase the block indent level and start a new line
  formatOpeningParentheses(tokens, index, query) {
    /*
         * Take out the preceding space unless there was whitespace there in the original query
         * or another opening parens
         **/
    const previousToken = tokens[index - 1];
    if (
      previousToken
            && previousToken.type !== tokenTypes.WHITESPACE
            && previousToken.type !== tokenTypes.OPEN_PAREN
    ) {
      query = trimEnd(query);
    }
    query += tokens[index].value;

    this.inlineBlock.beginIfPossible(tokens, index);

    if (!this.inlineBlock.isActive()) {
      this.indentation.increaseBlockLevel();
      query = this.addNewline(query);
    }
    return query;
  }

  // Closing parentheses decrease the block indent level
  formatClosingParentheses(token, query) {
    if (this.inlineBlock.isActive()) {
      this.inlineBlock.end();
      return this.formatWithSpaceAfter(token, query);
    }
    this.indentation.decreaseBlockLevel();
    return this.formatWithSpaces(token, this.addNewline(query));
  }

  formatPlaceholder(token, query) {
    return `${query + this.params.get(token)} `;
  }

  // Commas start a new line (unless within inline parentheses or SQL "LIMIT" clause)
  formatComma(token, query) {
    query = `${trimEnd(query) + token.value} `;

    if (this.inlineBlock.isActive()) {
      return query;
    }
    if (/^LIMIT$/i.test(this.previousReservedWord.value)) {
      return query;
    }
    return this.addNewline(query);
  }

  formatWithSpaceAfter(token, query) {
    return `${trimEnd(query) + token.value} `;
  }

  formatWithoutSpaces(token, query) {
    return trimEnd(query) + token.value;
  }

  formatWithSpaces(token, query) {
    return `${query + token.value} `;
  }

  addNewline(query) {
    return `${trimEnd(query)}\n${this.indentation.getIndent()}`;
  }
}
